docs/developer-docs/backend/rust/1-infrastructure.md

```diff
@@ -0,0 +1,204 @@
+
+#### Testing upgrade hooks
+
+It is best practice to test upgrades before applying them in order to catch any potential errors that may result in losing data irrevocably. To test upgrades, several different workflows or approaches can be used, such as shell or bash scripts, or Rust test scripts. The following psuedo-code showcases a Rust upgrade example that adds an additional step to execute the state validation of your upgrade test. 
+
+```
+let canister_id = install_canister(WASM);
+populate_data(canister_id);
+if should_upgrade { upgrade_canister(canister_id, WASM); }
+let data = query_canister(canister_id);
+assert_eq!(data, expected_value);
+```
+
+Then, your tests should be run twice in two different scenarios:
+- In a scenario without any upgrades, to assure that your tests run successfully without executing an upgrade.
+- In a scenario with an upgrade, to assure that your tests run successfully while executing an upgrade. 
+You then run your tests twice in different modes:
+
+By running both of these tests, developers can gain confidence that when an upgrade is applied to a canister, the canister's state is preserved. 
+
+:::caution
+It is not recommended to trap within the `pre_upgrade` hook. This is because while the `pre_upgrade` and `post_upgrade` hooks appear to be symmetrical, they are not. 
+
+If the `pre_upgrade` hook succeeds, but the `post_upgrade` hook traps, the canister can be debugged and another version can be built. However, if the `pre_upgrade` hook traps, there is not much you can do about it; a broken `pre_upgrade` hook prevents you from changing the canister's behavior. 
+:::
+
+#### Using stable memory as primary storage
+
+When a canister is upgraded, there is a limit regarding how many cycles a canister can burn during that upgrade. If the canister goes beyond that limit, the upgrade will be canceled by the system and the canister's state will be reverted. That means if you serialize the canister's whole state to stable memory in the `pre_upgrade` hook and the state becomes very large, the canister may not be able to be upgraded again. 
+
+One way to prevent this is to avoid serializing the canister state to begin with. Stable memory can be used as the canister's primary storage, where it can be used to store each upgrade call. Using this method, the `pre_upgrade` hook may not be necessary, and the `post_upgrade` hook will burn fewer cycles. 
```

Comment:
```suggestion
One way to prevent this is to avoid serializing the canister state to begin with. Stable memory can be used as the canister's primary storage, where it can be used to store modifications of each update call. Using this method, the `pre_upgrade` hook may not be necessary, and the `post_upgrade` hook will burn fewer cycles. 
```